src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,706 @@
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    // Portions of the code implemented below are based on the 'Berkeley SoftFloat Release 3e' algorithms.
+
+    /// <summary>
+    /// An IEEE 754 compliant float16 type.
+    /// </summary>
+    [Serializable]
+    [StructLayout(LayoutKind.Sequential)]
+    public readonly struct Half : IComparable, IFormattable, IComparable<Half>, IEquatable<Half>, ISpanFormattable
+    {
+        private const NumberStyles DefaultParseStyle = NumberStyles.Float | NumberStyles.AllowThousands;
+
+        // Constants for manipulating the private bit-representation
+
+        private const ushort SignMask = 0x8000;
+        private const ushort SignShift = 15;
+
+        private const ushort ExponentMask = 0x7C00;
+        private const ushort ExponentShift = 10;
+
+        private const ushort SignificandMask = 0x03FF;
+        private const ushort SignificandShift = 0;
+
+        private const ushort MinSign = 0;
+        private const ushort MaxSign = 1;
+
+        private const ushort MinExponent = 0x00;
+        private const ushort MaxExponent = 0x1F;
+
+        private const ushort MinSignificand = 0x0000;
+        private const ushort MaxSignificand = 0x03FF;
+
+        // Constants representing the private bit-representation for various default values
+
+        private const ushort PositiveZeroBits = 0x0000;
+        private const ushort NegativeZeroBits = 0x8000;
+
+        private const ushort EpsilonBits = 0x0001;
+
+        private const ushort PositiveInfinityBits = 0x7C00;
+        private const ushort NegativeInfinityBits = 0xFC00;
+
+        private const ushort PositiveQNaNBits = 0x7E00;
+        private const ushort NegativeQNaNBits = 0xFE00;
+
+        private const ushort MinValueBits = 0xFBFF;
+        private const ushort MaxValueBits = 0x7BFF;
+
+
+        //
+        // Well-defined and commonly used values
+        //
+
+        public static readonly Half Epsilon = new Half(EpsilonBits);                        //  5.9604645E-08
+
+        public static readonly Half PositiveInfinity = new Half(PositiveInfinityBits);      //  1.0 / 0.0
+        public static readonly Half NegativeInfinity = new Half(NegativeInfinityBits);      // -1.0 / 0.0
+
+        public static readonly Half NaN = new Half(NegativeQNaNBits);                       //  0.0 / 0.0
+
+        public static readonly Half MinValue = new Half(MinValueBits);                      // -65504
+        public static readonly Half MaxValue = new Half(MaxValueBits);                      //  65504
+
+        // We use these explicit definitions to avoid the confusion between 0.0 and -0.0.
+        private static readonly Half PositiveZero = new Half(PositiveZeroBits);            //  0.0
+        private static readonly Half NegativeZero = new Half(NegativeZeroBits);            // -0.0
+
+        private readonly ushort m_value; // Do not rename (binary serialization)
```

Comment:
Although my care level is low, to be consistent this should have an entry in https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs around line 116.

Reply:
I should not have said low - either we support binary-serializing this or we don't, and if we do it needs a test there.

Reply:
@bartonjs, @joperezr (and maybe @GrabYourPitchforks), what are your recommendations here?

This comment was copied because its on the other primitive types, but IIRC binary serialization is mostly legacy now, so this might not be worth supporting on a new type?

Reply:
Binary serialization is legacy. We should not be adding binary serialization to any new types.

Reply:
Sounds good to me.

Reply:
Done. Removed both the attribute and comment